### DISCUSSION TITLE
Using the same regex as `kitty @ kitten hints --type linenum` in a custom hint?

### DISCUSSION THREAD

**Discussion by cpdean** [kovidgoyal/kitty#6393] [Q&A] [answered]
I'd like to make a custom hint so that I can use the same keybinding for opening files at a specific line number between languages. The built-in hint `linenum` works with Rust (and grep, and many other things), to parse the path and line number out:

```
at ./src/main.rs:3:5
```

However, I want to use the same command to parse both rust and python stack traces.
```
File "/Users/c/.dotfiles/what.py", line 8, in bar
```

I can get Python stack traces working with the following:

```
kitty @ kitten hints --type=linenum --regex='File "(?P<path>[^"]+)", line (?P<line>\\d+),'
```

Now I was hoping I could have the same hint work for both formats by combining the regexes, something like

```
kitty @ kitten hints --type=linenum --regex="$DEFAULT_LINENUM_REGEX|$PYTHON_TRACEBACK"
```

However, the regex I find in the [source code](https://github.com/kovidgoyal/kitty/blob/07bcc5ba61260a6953121f63416c3cfc12cadb46/kittens/hints/marks.go#L27-L35) does not appear to have the same behavior as in Python's regex dialect. I'm testing it by copying it over to Python and testing it the same way I did to write the my custom matcher above.

```python
import re

FILE_EXTENSION = r"\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?:\b|[^.])"
path_regex = r"(?:\S*?/[\r\S]+)|(?:\S[\r\S]*%s)\b" % FILE_EXTENSION
default_linenum_regex = r"(?P<path>%s):(?P<line>\d+)" % path_regex

line = "             at ./src/main.rs:3:5"
re.search(default_linenum_regex, line).groupdict()
# {'path': './src/main.rs:3', 'line': '5'}
```

Instead of getting the line number 3, it considers `.rs:3` to be the file extension and parses the column number as the line number.

kitty version:

```
kitty --version
kitty 0.28.1 created by Kovid Goyal
```

How do I extend the default linenum hint with my additional regex for Python?

Thanks!

---
### testing details

This is what I've been using to test regexes. This is formatted in a way that it should be copy-pastable to a python3 interpreter.

```python
rust_stack = """
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/main.rs:3:7
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:117:5
   3: core::option::Option<T>::unwrap
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/option.rs:950:21
   4: backtrace_test::foo
             at ./src/main.rs:3:5
   5: backtrace_test::bar
             at ./src/main.rs:7:5
   6: backtrace_test::main
             at ./src/main.rs:12:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
""".split("\n")

python_stack = """
Traceback (most recent call last):
  File "/Users/c/.dotfiles/what.py", line 10, in <module>
    bar()
  File "/Users/c/.dotfiles/what.py", line 8, in bar
    return goat()
           ^^^^^^
  File "/Users/c/.dotfiles/what.py", line 5, in goat
    return foo()
           ^^^^^
  File "/Users/c/.dotfiles/what.py", line 2, in foo
    raise Exception("boop")
Exception: boop""".split("\n")

rust_line = rust_stack[12]
python_line = python_stack[7]

import re

python_traceback = 'File "(?P<path>[^"]+)", line (?P<line>\d+),'

FILE_EXTENSION = r"\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?:\b|[^.])"
path_regex = r"(?:\S*?/[\r\S]+)|(?:\S[\r\S]*%s)\b" % FILE_EXTENSION
default_linenum_regex = r"(?P<path>%s):(?P<line>\d+)" % path_regex

re.search(python_traceback, python_line).groupdict()
# {'path': '/Users/c/.dotfiles/what.py', 'line': '5'}

re.search(default_linenum_regex, rust_line).groupdict()
# {'path': './src/main.rs:3', 'line': '5'}


# edit: found the old python impl at @kovidgoyal's suggestion
# git log -p -S"P<path>" | grep -A4 FILE_EXT | head -n4

og_FILE_EXTENSION = r'\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?!\.)'
og_PATH_REGEX = fr'(?:\S*?/[\r\S]+)|(?:\S[\r\S]*{og_FILE_EXTENSION})\b'
og_DEFAULT_LINENUM_REGEX = fr'(?P<path>{og_PATH_REGEX}):(?P<line>\d+)'

re.search(og_DEFAULT_LINENUM_REGEX, rust_line).groupdict()
# {'path': './src/main.rs:3', 'line': '5'}

```

**Comment by kovidgoyal** [comment]
Look in the git history from before this kitten was ported to Go and you
will find the python version of the regex.

**Reply by cpdean** [reply]
@kovidgoyal that's a good idea

I found the [old implementation](https://github.com/kovidgoyal/kitty/blob/2e1eebd9986a5ec9a69dd0ff534a35d342a78a73/kittens/hints/main.py#L39-L42) now with this

```
git log -p -S"P<path>" | grep -A4 FILE_EXT | head -n4
```


When testing it in a Python interperter, it has the same behavior of the capture group for the file path greedily grabbing the line number.

```
og_FILE_EXTENSION = r'\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?!\.)'
og_PATH_REGEX = fr'(?:\S*?/[\r\S]+)|(?:\S[\r\S]*{og_FILE_EXTENSION})\b'
og_DEFAULT_LINENUM_REGEX = fr'(?P<path>{og_PATH_REGEX}):(?P<line>\d+)'

re.search(og_DEFAULT_LINENUM_REGEX, rust_line).groupdict()
# {'path': './src/main.rs:3', 'line': '5'}
```

I decided to stick it in my custom command anyway to see if there's something else I'm getting wrong about how this flows through kitty, and I'm getting a syntax error:

```
kitty @ kitten hints --type=linenum --regex='File "(?P<path>[^"]+)", line (?P<line>\\d+),|(?P<path>(?:\S*?/[\r\S]+)|(?:\S[\r\S]*\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?!\.))\b):(?P<line>\d+)'


Error: Failed to compile the regex pattern: "File \"(?P<path>[^\"]+)\", line (?P<line>\\d+),|(?P<path>(?:\\S*?/[\\r\\S]+)|(?:\\S[\\r\\S]*\\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?!\\.))\\b):(?P<line>\\d+)" with error: error parsing regexp: invalid or unsupported Perl syntax: `(?!`
```

This is odd, because that had worked in a Python interpreter. Maybe this goes through the go runtime now and the docs have gotten out of date?

> The regular expression to use when option [--type](https://sw.kovidgoyal.net/kitty/kittens/hints/#cmdoption-kitty-kitten-hints-type) is set to regex, in python syntax.

[--regex flag](https://sw.kovidgoyal.net/kitty/kittens/hints/#cmdoption-kitty-kitten-hints-regex)

If that's the case, then I should not use Python to validate these, so trying the regex from `kittens/hints/marks.go` in the command and it now matches both Python and Rust stack traces!

```
kitty @ kitten hints --type=linenum --regex='File "(?P<path>[^"]+)", line (?P<line>\\d+),|(?P<path>(?:\S*?/[\r\S]+)|(?:\S[\r\S]*\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?:\b|[^.]))\b):(?P<line>\d+)'
```

**Comment by cpdean** [comment, marked as answer]
After fixing some shell escaping I now have replaced the default behavior and mapping with one that can work with both line number styles:

```
map kitty_mod+p>n kitten hints --type=linenum --regex="(?:File \"(?P<path>[^\"]+)\", line (?P<line>\\d+),)|(?P<path>(?:\S*?/[\r\S]+)|(?:\S[\r\S]*\.(?:[a-zA-Z0-9]{2,7}|[ahcmo])(?:\b|[^.]))\b):(?P<line>\d+)"
```
